junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ResourceLock.java

```diff
@@ -43,12 +45,18 @@ default void close() {
 		release();
 	}
 
+	SortedSet<ExclusiveResource> getResources();
+
 	/**
 	 * {@return whether the given lock is compatible with this lock}
 	 * @param other the other lock to check for compatibility
 	 */
 	default boolean isCompatible(ResourceLock other) {
-		return this instanceof NopLock || other instanceof NopLock;
+		if (this.getResources().isEmpty() || other.getResources().isEmpty()) {
+			return true;
+		}
+		return other.getResources().stream() //
+				.allMatch(it -> this.getResources().contains(it)
+						|| ExclusiveResource.COMPARATOR.compare(this.getResources().last(), it) < 0);
```

Comment:
Can't this be simplified to
```suggestion
				.allMatch(it -> ExclusiveResource.COMPARATOR.compare(this.getResources().last(), it) <= 0);
```

Reply:
No, since that would disallow "common prefixes", e.g. both having the global read lock but `other` having another (read) lock.